vizro-core/tests/unit/vizro/models/_components/test_aggrid.py

```diff
@@ -0,0 +1,156 @@
+from vizro.managers import data_manager
+from vizro.models._action._action import Action
+from vizro.tables import dash_ag_grid
+
+
+@pytest.fixture
+def dash_ag_grid_with_arguments():
+    return dash_ag_grid(data_frame=px.data.gapminder(), defaultColDef={"resizable": False, "sortable": False})
+
+
+@pytest.fixture
+def dash_ag_grid_with_str_dataframe():
+    return dash_ag_grid(data_frame="gapminder")
+
+
+class TestDunderMethodsAgGrid:
+    def test_create_graph_mandatory_only(self, standard_ag_grid):
+        ag_grid = vm.AgGrid(figure=standard_ag_grid)
+
+        assert hasattr(ag_grid, "id")
+        assert ag_grid.type == "ag_grid"
+        assert ag_grid.figure == standard_ag_grid
+        assert ag_grid.actions == []
+
+    @pytest.mark.parametrize("id", ["id_1", "id_2"])
+    def test_create_ag_grid_mandatory_and_optional(self, standard_ag_grid, id):
+        ag_grid = vm.AgGrid(figure=standard_ag_grid, id=id, actions=[])
+
+        assert ag_grid.id == id
+        assert ag_grid.type == "ag_grid"
+        assert ag_grid.figure == standard_ag_grid
+
+    def test_mandatory_figure_missing(self):
+        with pytest.raises(ValidationError, match="field required"):
+            vm.AgGrid()
+
+    def test_failed_ag_grid_with_no_captured_callable(self, standard_go_chart):
+        with pytest.raises(ValidationError, match="must provide a valid CapturedCallable object"):
+            vm.AgGrid(figure=standard_go_chart)
+
+    @pytest.mark.xfail(reason="This test is failing as we are not yet detecting different types of captured callables")
+    def test_failed_ag_grid_with_wrong_captured_callable(self, standard_px_chart):
+        with pytest.raises(ValidationError, match="must provide a valid ag_grid function vm.AgGrid"):
+            vm.AgGrid(figure=standard_px_chart)
+
+    def test_getitem_known_args(self, dash_ag_grid_with_arguments):
+        ag_grid = vm.AgGrid(figure=dash_ag_grid_with_arguments)
+        assert ag_grid["defaultColDef"] == {"resizable": False, "sortable": False}
+        assert ag_grid["type"] == "ag_grid"
+
+    def test_getitem_unknown_args(self, standard_ag_grid):
+        ag_grid = vm.AgGrid(figure=standard_ag_grid)
+        with pytest.raises(KeyError):
+            ag_grid["unknown_args"]
+
+    def test_set_action_via_validator(self, standard_ag_grid, identity_action_function):
+        ag_grid = vm.AgGrid(figure=standard_ag_grid, actions=[Action(function=identity_action_function())])
+        actions_chain = ag_grid.actions[0]
+        assert actions_chain.trigger.component_property == "cellClicked"
+
+
+class TestAttributesAgGrid:
+    def test_ag_grid_filter_interaction_attributes(self, ag_grid_with_id):
+        ag_grid = vm.AgGrid(figure=ag_grid_with_id, title="Gapminder", actions=[])
+        assert hasattr(ag_grid, "_filter_interaction")
+        ag_grid.pre_build()
+        assert hasattr(ag_grid, "_filter_interaction_input")
+        assert "modelID" in ag_grid._filter_interaction_input
+
```

Comment:
```suggestion
```

This is too much of an implementation detail to test. I assume that if these didn't exist then some other actions tests would fail? If so then let's remove this one and the similar ones in the graph and table tests.

Reply:
It's a bit tricky - I created them for the precise reason that other tests are not failing (at least in the current design)!

The reason for this is that we mock the callback context when "running" actions - so we fake the user clicking on something. If we did not do that the testing of actions would be a lot more complicated

You are correct in that the `assert hasattr(ag_grid, "_filter_interaction_input")` is unncessary in that sense, but I added it for completeness.

I would propose to remove that one then, and leave a comment why this is important

Reply:
Ahhh ok, if the others tests aren't failing then we should definitely leave this here. I'm fine to leave all the `hasattr` there.

The fact that the actions tests _don't_ fail if you remove this property indicates a deficiency with our testing of actions, but given this implementation may well change anyway let's solve that another time as part of the actions v2 work @petar-qb.

Reply:
Yes, so just FYI when we come back to this, so that we can remember: this is part of the build method of the `Action` model, so it will only get triggered then.

Currently we do not build the `Action` model in order to test the functionality - presumably because we fake the ctx